cmd/argocd/commands/cluster.go

```diff
@@ -147,9 +152,18 @@ func NewClusterAddCommand(clientOpts *argocdclient.ClientOptions, pathOpts *clie
 				contextName = clusterOpts.Name
 			}
 			clst := cmdutil.NewCluster(contextName, clusterOpts.Namespaces, clusterOpts.ClusterResources, conf, managerBearerToken, awsAuthConf, execProviderConf, labelsMap, annotationsMap)
-			if clusterOpts.InCluster {
+			if clusterOpts.InCluster || clusterOpts.ClusterEndpoint == string(cmdutil.KubeInternalEndpoint) {
 				clst.Server = argoappv1.KubernetesInternalAPIServerAddr
+			} else if clusterOpts.ClusterEndpoint == string(cmdutil.KubePublicEndpoint) {
+				endpoint, err := cmdutil.GetKubePublicEndpoint(clientset)
+				if err != nil || len(endpoint) == 0 {
+					log.Warn("Failed to find the cluster endpoint from kube-public data. Falling back to the endpoint listed in the kubconfig context.")
+					log.Debugf("Get kube-public endpoint error: %v", err)
```

Comment:
I think we may want to print the error together with the warning, and not only when in debug mode. I'd suggest the either the following:

```suggestion
					log.Warnf("Failed to find the cluster endpoint from kube-public data: %v", err)
					log.Infof("Falling back to the endpoint '%s' as listed in the kubeconfig context", clst.Server)
```

I'd also like to raise the question whether it's in the intention of the user that a fallback is used when the requested method fails.

Reply:
I added fallback with the hypothesis that the endpoint in `kube-public` is usually the same as the one in the user's kubeconfig. You brought up k3s in the feature - it's been a minute since I've used that distro, so I'm not sure what it publishes there.

I have the fallback because the other methods won't fail resolving the URL.